### PR TITLE
Add one option to remove extra residual evaluation if a linear solver is chosen

### LINKED ISSUE #11760: perform only 1 residual calculation if solve_type=Linear
## Rationale

Only 1 residual calculation per timestep is needed if solve_type=Linear.

## Description

At the moment two residual calculations are performed every timestep.  One for the first nonlinear residual (this is required), then PETSc does its linear solve, and then another, redundant, nonlinear residual calculation.  I do not know if it is PETSc, libmesh or MOOSE that is requesting this second residual calculation.  

## Impact

Computational efficiency.  In the simple physics i happen to be modelling right now, just doing 1 residual calculation would speed up MOOSE by two times!

### PR DESCRIPTION
**Pull request by fdkong**
Closes #11760 #11761

### CODE REVIEW

File: test/tests/kernels/2d_diffusion/tests [hunk 1]
```diff
@@ -36,4 +36,11 @@
     petsc_version = '>=3.8.4'
     prereq = testdirichlet
   [../]
+
+  [./actual_linear_solver]
+    type = RunApp
+    input = '2d_diffusion_test.i'
+    cli_args = 'Executioner/solve_type=linear Outputs/exodus=false'
+    absent_out = 'Nonlinear \|R\|'
```
Comment (lindsayad) [inline test/tests/kernels/2d_diffusion/tests]: You need `issues`, `design`, and `requirement` 😄

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Comment by WilkAndy** [comment]
Oh, this does not work!  There is no output at "1 Nonlinear |R|", so it looks like the residual's only being computed once.  However, MOOSE still does compute the residual twice.  You can check that by ```print_perf_log = true``` in your test file.

**Comment by lindsayad** [comment]
Not `--dash-beautiful-code`?

**Comment by lindsayad** [comment]
@WilkAndy This works for me. Without @fdkong change to `PetscSupport.C`:
```
 -------------------------------------------------------------------------------------------------------------------------
| Moose Test Performance: Alive time=0.045687, Active time=0.034562                                                       |
 -------------------------------------------------------------------------------------------------------------------------
| Event                                      nCalls     Total Time  Avg Time    Total Time  Avg Time    % of Active Time  |
|                                                       w/o Sub     w/o Sub     With Sub    With Sub    w/o S    With S   |
|-------------------------------------------------------------------------------------------------------------------------|
|                                                                                                                         |
|                                                                                                                         |
| Application                                                                                                             |
|   Full Runtime                             1          0.0005      0.000463    0.0346      0.034563    1.34     100.00   |
|                                                                                                                         |
| Execution                                                                                                               |
|   computeKernels()                         2          0.0001      0.000047    0.0001      0.000047    0.27     0.27     |
|   computeNodalBCs()                        2          0.0000      0.000004    0.0000      0.000004    0.02     0.02     |
|   compute_jacobian()                       1          0.0001      0.000128    0.0001      0.000128    0.37     0.37     |
|   compute_residual()                       2          0.0002      0.000112    0.0003      0.000163    0.65     0.95     |
|   solve()                                  1          0.0014      0.001387    0.0018      0.001842    4.01     5.33     |
```
With his change:
```
 -------------------------------------------------------------------------------------------------------------------------
| Moose Test Performance: Alive time=0.044943, Active time=0.034394                                                       |
 -------------------------------------------------------------------------------------------------------------------------
| Event                                      nCalls     Total Time  Avg Time    Total Time  Avg Time    % of Active Time  |
|                                                       w/o Sub     w/o Sub     With Sub    With Sub    w/o S    With S   |
|-------------------------------------------------------------------------------------------------------------------------|
|                                                                                                                         |
|                                                                                                                         |
| Application                                                                                                             |
|   Full Runtime                             1          0.0004      0.000448    0.0344      0.034395    1.30     100.00   |
|                                                                                                                         |
| Execution                                                                                                               |
|   computeKernels()                         1          0.0001      0.000060    0.0001      0.000060    0.17     0.17     |
|   computeNodalBCs()                        1          0.0000      0.000006    0.0000      0.000006    0.02     0.02     |
|   compute_jacobian()                       1          0.0001      0.000093    0.0001      0.000093    0.27     0.27     |
|   compute_residual()                       1          0.0001      0.000108    0.0002      0.000174    0.31     0.51     |
|   solve()                                  1          0.0012      0.001247    0.0015      0.001514    3.63     4.40     |
```
If I build a debug executable and break on `NonlinearSystemBase::computeResidualTags`, I see that it is indeed called only once. Now I am on Petsc 3.9.2...let me check on older Petsc.

**Comment by lindsayad** [comment]
Ok, I'm seeing the two calls in 3.7.6

**Comment by lindsayad** [comment]
Yep, behavior changed in `ksponly.c` from PETSc versions. In 3.7.6:
```
  /* Take the computed step. */
  ierr = VecAXPY(X,-1.0,Y);CHKERRQ(ierr);
  ierr = SNESComputeFunction(snes,X,F);CHKERRQ(ierr);
  if (snes->numbermonitors) {
    PetscReal fnorm;
    ierr = VecNorm(F,NORM_2,&fnorm);CHKERRQ(ierr);
    ierr = SNESMonitor(snes,1,fnorm);CHKERRQ(ierr);
  }
  PetscFunctionReturn(0);
```
Starting in 3.8 (and actually before 3.6.1), this was changed to:
```
  /* Take the computed step. */
  ierr = VecAXPY(X,-1.0,Y);CHKERRQ(ierr);
  if (snes->numbermonitors) {
    PetscReal fnorm;
    ierr = SNESComputeFunction(snes,X,F);CHKERRQ(ierr);
    ierr = VecNorm(F,NORM_2,&fnorm);CHKERRQ(ierr);
    ierr = SNESMonitor(snes,1,fnorm);CHKERRQ(ierr);
  }
  PetscFunctionReturn(0);
```
Motivation for upgrading PETSc

**Review by lindsayad — APPROVED** [review]

**Comment by WilkAndy** [comment]
OK.  Thanks @lindsayad .  So petsc3.8 seems to be the best, correct?  i didn't think MOOSE worked with 3.9 yet.   This is rather "delicate"!  How do we know that 3.10 won't change again?!

**Comment by fdkong** [comment]
Right, I should mention the PETSc version. I Fixed this issue in PETSc-3.8.x: if there is no monitor, then we do not have any reason to call computeResidual.

**Comment by lindsayad** [comment]
The only things that don't work with petsc 3.9 are the Jacobian testers and we have a PR up for that right now.

**Comment by joshuahansel** [comment]
By the way, I am attempting to investigate the RELAP-7 failures. I've pretty much been compiling all day, which takes a while off-site (and of course, I forgot that I needed to update my MOOSE package to get the new PETSc option). Anyway, I think to put out this fire for now, I'm going to skip these tests with a new RELAP-7 issue to buy me time to figure things out. The failures seem to suggest that some object isn't clearing some data before the unnecessary 2nd residual evaluation. I suspect that some base classes in the RDG module are to blame, but we'll see.

**Comment by joshuahansel** [comment]
I have looked into the RELAP-7 failures, and I believe I have determined the issue. It appears that there is an output bug for `solve_type=LINEAR` in MOOSE associated with the new change: aux variables are lagged by a time step.

I did the following:
* reproduced the exodiff with the new change
* verified that the residuals are exactly the same before and after the change
* verified that the Jacobians are exactly the same before and after the change
* plotted solution variables before and after the change against each other and verified they were the same
* plotted aux variables before and after the change against each other and verified they were lagged with the new change

It seems as if the aux variable values were being updated in that SNES monitor evaluation with the new end-of-time-step solution, and now they are not. I have not yet looked into why.

**Comment by lindsayad** [comment]
What's the `execute_on` for those auxiliary variables?

**Comment by joshuahansel** [comment]
`initial timestep_end`

**Comment by joshuahansel** [comment]
See #11798 for an example of the issue.

**Pull request merged**